persistence/src/commonMain/kotlin/com/wire/kalium/persistence/dao/message/MessageDAO.kt

```diff
@@ -3,27 +3,89 @@ package com.wire.kalium.persistence.dao.message
+    open val assetAudioDurationMs: Long? = null,
+    open val assetAudioNormalizedLoudness: ByteArray? = null,
+    open val assetOtrKey: ByteArray? = null,
+    open val assetSha256Key: ByteArray? = null,
+    open val assetId: String? = null,
+    open val assetToken: String? = null,
+    open val assetDomain: String? = null,
+    open val assetEncryptionAlgorithm: String? = null,
 ) {
     enum class Status {
         PENDING, SENT, READ, FAILED
     }
+
+    data class TextMessageEntity(
+        // Text message specific field
+        override val content: String?,
+
+        // BaseMessageEntity fields
+        override val id: String,
+        override val conversationId: QualifiedIDEntity,
+        override val date: String,
+        override val senderUserId: QualifiedIDEntity,
+        override val senderClientId: String,
+        override val status: Status
+    ) : BaseMessageEntity(id, conversationId, date, senderUserId, senderClientId, status, content)
+
+    data class AssetMessageEntity(
+        // Asset message specific fields
+        override val assetMimeType: String? = null,
+        override val assetSize: Int? = null,
+        override val assetName: String? = null,
+        override val assetImageWidth: Int? = null,
+        override val assetImageHeight: Int? = null,
+        override val assetVideoWidth: Int? = null,
+        override val assetVideoHeight: Int? = null,
+        override val assetVideoDurationMs: Long? = null,
+        override val assetAudioDurationMs: Long? = null,
+        override val assetAudioNormalizedLoudness: ByteArray? = null,
+        override val assetOtrKey: ByteArray? = null,
+        override val assetSha256Key: ByteArray? = null,
+        override val assetId: String? = null,
+        override val assetToken: String? = null,
+        override val assetDomain: String? = null,
+        override val assetEncryptionAlgorithm: String? = null,
+
+        // BaseMessageEntity fields
+        override val id: String,
+        override val conversationId: QualifiedIDEntity,
+        override val date: String,
+        override val senderUserId: QualifiedIDEntity,
+        override val senderClientId: String,
+        override val status: Status
+    ) : BaseMessageEntity(id, conversationId, date, senderUserId, senderClientId, status)
 }
 
 interface MessageDAO {
     suspend fun deleteMessage(id: String, conversationsId: QualifiedIDEntity)
     suspend fun deleteAllMessages()
-    suspend fun insertMessage(message: MessageEntity)
-    suspend fun insertMessages(messages: List<MessageEntity>)
-    suspend fun updateMessage(message: MessageEntity)
-    suspend fun getAllMessages(): Flow<List<MessageEntity>>
-    suspend fun getMessageById(id: String, conversationId: QualifiedIDEntity): Flow<MessageEntity?>
-    suspend fun getMessageByConversation(conversationId: QualifiedIDEntity, limit: Int): Flow<List<MessageEntity>>
+    suspend fun insertMessage(message: BaseMessageEntity)
```

Comment:
How about enforcing correct types by adding an `insert` per type.

```kotlin
suspend fun insertTextMessage(message: TextMessageEntity)
suspend fun insertAssetMessage(message: AssetMessageEntity)
```

Reply:
Or maybe we can do a similar thing like you do in the logic data model and use a sealed class.

```kotlin
data class MessageEntity {
   val id: String,
   ...
   ...
   val content: MessageContent
}

sealed class MessageContent {
   data class TextContent {}
   data class AssetContent {}
}